pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -1356,6 +1342,58 @@ protected void handleGetSchema(CommandGetSchema commandGetSchema) {
         });
     }
 
+    @Override
+    protected void handleGetOrCreateSchema(CommandGetOrCreateSchema commandGetOrCreateSchema) {
+        if (log.isDebugEnabled()) {
+            log.debug("Received CommandGetOrCreateSchema call from {}", remoteAddress);
+        }
+        long requestId = commandGetOrCreateSchema.getRequestId();
+        String topicName = commandGetOrCreateSchema.getTopic();
+        SchemaData schemaData = getSchema(commandGetOrCreateSchema.getSchema());
+        SchemaData schema = schemaData.getType() == SchemaType.NONE ? null : schemaData;
+        service.getTopicIfExists(topicName).thenAccept(topicOpt -> {
+            if (topicOpt.isPresent()) {
+                Topic topic = topicOpt.get();
+                CompletableFuture<SchemaVersion> schemaVersionFuture = tryAddSchema(topic, schema);
+                schemaVersionFuture.exceptionally(ex -> {
+                    ServerError errorCode = BrokerServiceException.getClientErrorCode(ex);
+                    ctx.writeAndFlush(Commands.newGetOrCreateSchemaResponseError(
+                            requestId, errorCode, ex.getMessage()));
+                    return null;
+                }).thenAccept(schemaVersion -> {
+                        ctx.writeAndFlush(Commands.newGetOrCreateSchemaResponse(
+                                requestId, schemaVersion));
+                });
+            } else {
+                ctx.writeAndFlush(Commands.newGetOrCreateSchemaResponseError(
+                        requestId, ServerError.TopicNotFound, "Topic not found"));
+            }
+        }).exceptionally(ex -> {
+            ServerError errorCode = BrokerServiceException.getClientErrorCode(ex);
+            ctx.writeAndFlush(Commands.newGetOrCreateSchemaResponseError(
+                    requestId, errorCode, ex.getMessage()));
+            return null;
+        });
+    }
+
+    private CompletableFuture<SchemaVersion> tryAddSchema(Topic topic, SchemaData schema) {
+        if (schema != null) {
+            return topic.addSchema(schema);
+        } else {
+            return topic.hasSchema().thenCompose((hasSchema) -> {
+                log.info("[{}] {} configured with schema {}",
```

Comment:
Can we keep the origin log statement when refactoring the code?

Reply:
`producer_id` not carried in `CommandGetOrCreateSchema`, actually need it? 
Or a `-1` as placeholder?